examples/cjs.py

```diff
@@ -0,0 +1,288 @@
+
+We show how to implement several variants of the Cormack-Jolly-Seber (CJS)
+[4, 5, 6] model used in ecology to analyze animal capture-recapture data.
+For a discussion of these models see reference [1].
+
+We make use of two datasets:
+-- the European Dipper (Cinclus cinclus) data from reference [2]
+   (this is Norway's national bird).
+-- the meadow voles data from reference [3].
+
+Compare to the Stan implementations in [7].
+
+References
+[1] Kery, M., & Schaub, M. (2011). Bayesian population analysis using
+    WinBUGS: a hierarchical perspective. Academic Press.
+[2] Lebreton, J.D., Burnham, K.P., Clobert, J., & Anderson, D.R. (1992).
+    Modeling survival and testing biological hypotheses using marked animals:
+    a unified approach with case studies. Ecological monographs, 62(1), 67-118.
+[3] Nichols, Pollock, Hines (1984) The use of a robust capture-recapture design
+    in small mammal population studies: A field example with Microtus pennsylvanicus.
+    Acta Theriologica 29:357-365.
+[4] Cormack, R.M., 1964. Estimates of survival from the sighting of marked animals.
+    Biometrika 51, 429-438.
+[5] Jolly, G.M., 1965. Explicit estimates from capture-recapture data with both death
+    and immigration-stochastic model. Biometrika 52, 225-247.
+[6] Seber, G.A.F., 1965. A note on the multiple recapture census. Biometrika 52, 249-259.
+[7] https://github.com/stan-dev/example-models/tree/master/BPA/Ch.07
+[8] http://pyro.ai/examples/capture_recapture.html
+"""
+
+import argparse
+import os
+
+from jax import random
+import jax.numpy as jnp
+from jax.scipy.special import expit, logit
+
+import numpyro
+from numpyro import handlers
+from numpyro.contrib.control_flow import scan
+import numpyro.distributions as dist
+from numpyro.examples.datasets import DIPPER_VOLE, load_dataset
+from numpyro.infer import HMC, MCMC, NUTS
+from numpyro.infer.reparam import LocScaleReparam
+
+
+# %%
+# Our first and simplest CJS model variant only has two continuous
+# (scalar) latent random variables: i) the survival probability phi;
+# and ii) the recapture probability rho. These are treated as fixed
+# effects with no temporal or individual/group variation.
+
+
+def model_1(capture_history, sex):
+    N, T = capture_history.shape
+    phi = numpyro.sample("phi", dist.Uniform(0.0, 1.0))  # survival probability
+    rho = numpyro.sample("rho", dist.Uniform(0.0, 1.0))  # recapture probability
+
+    def transition_fn(carry, y):
+        first_capture_mask, z = carry
+        with numpyro.plate("animals", N, dim=-1):
+            with handlers.mask(mask=first_capture_mask):
+                mu_z_t = first_capture_mask * phi * z + (1 - first_capture_mask)
+                # NumPyro exactly sums out the discrete states z_t.
+                z = numpyro.sample("z", dist.Bernoulli(dist.util.clamp_probs(mu_z_t)))
+                mu_y_t = rho * z
+                numpyro.sample("y", dist.Bernoulli(dist.util.clamp_probs(mu_y_t)), obs=y)
+
```

Comment:
does this treat `t=0` differently? in the pyro code `first_capture_mask` is all zeros for the first time slice?

Reply:
It is the same. I just unroll the first step. Could you double-check the logic?

Reply:
maybe i don't understand how numpyro's `scan` works but to me it looks like the `t=0` iteration is missing.
how do you get an observation statement with `obs=capture_history[:, 0]`?

Reply:
or is it just that that observation statement has zero `log_prob` anyway because of the mask?

Reply:
Yes, I moved the case t=0 out of the loop. At t=0, `mu_z_t=1` so `z=Bernoulli(mu_z_t)=1` (which is the same as the initial `z`) and `first_capture_mask = first_capture_mask | capture_history[:, 0] = capture_history[:, 0]`. Because the mask of first step is False, adding an obs statement is not necessary. We can add it but the code will be more populated:
```
first_capture_mask = False
with numpyro.plate("animals", N, dim=-1):
    with handlers.mask(mask=first_capture_mask):
        numpyro.sample("y", dist.Bernoulli(rho), obs=capture_history[:, 0])
first_capture_mask |= capture_history[:, 0].astype(bool)
```
comparing to 1 line of code
```
first_capture_mask = capture_history[:, 0].astype(bool)
```